src/main/scala/io/iohk/ethereum/mining/BlockGenerator.scala

```diff
@@ -37,11 +37,13 @@ class BlockGenerator(blockchainStorages: BlockchainStorages, blockchainConfig: B
         val blockTimestamp = blockTimestampProvider.getEpochSecond
         val header: BlockHeader = prepareHeader(blockNumber, ommers, beneficiary, parent, blockTimestamp)
 
-        val transactionsForBlock = transactions
-          .filter(_.tx.gasLimit < header.gasLimit)
-          //if we have 2 transactions from same address we want first one with lower nonce
-          .sortBy(_.tx.gasPrice)
-          .sortBy(_.tx.nonce)
+        val sortedTransactions = transactions.groupBy(_.senderAddress).values.toList.flatMap { txsFromSender =>
+          val ordered = txsFromSender.sortBy(_.tx.gasPrice).reverse.sortBy(_.tx.nonce)
```

Comment:
We already have the validation split, because of the gas limit check. What @ntallar proposes is rather an optimisation, we know this TX will fail, so why bother executing it? 

But I think the optimisation should go even further. Currently when we have a failing TX we start over, executing the block from scratch with the failing TX filtered out. But we should be able to revert to the state after previous TX and just continue from there. This definitely a separate task though.